File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<String,List<TreeNode>>map=new HashMap<>();
    public String dfs(TreeNode t){
        if(t==null)return "";
        String s="("+dfs(t.left)+t.val+dfs(t.right)+")";
        if(!map.containsKey(s))map.put(s,new ArrayList<TreeNode>());
        map.get(s).add(t);
        return s;
    }
    
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        dfs(root);
        List<TreeNode>res=new ArrayList<>();
        for(List<TreeNode>l:map.values()){
            if(l.size()>1)res.add(l.get(0));
        }
        return res;
    }
}

File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<string,vector<TreeNode*>>map;
    string dfs(TreeNode*t){
        if(!t)return "";
        string s="("+dfs(t->left)+to_string(t->val)+dfs(t->right)+")";
        map[s].push_back(t);
        return s;
    }
    
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        dfs(root);
        vector<TreeNode*>ans;
        
        for(auto s:map){
            if(s.second.size()>1)ans.push_back(s.second[0]);
        }
        return ans;
    }
};